### PR TITLE
satellite/satellitedb: separate uuid creation from db layer

### PR DESCRIPTION
**Pull request by VinozzZ**
What: 
- move uuid generation out of `Insert` call from users table

Why:
- referral manager needs access to the userID before it gets inserted into the satellitedb

Please describe the tests:
 - Test 1:
 - Test 2:
 
Please describe the performance impact:

## Code Review Checklist (to be filled out by reviewer)
 - [x] NEW: Are there any Satellite database migrations? Are they forwards _and_ backwards compatible? 
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

File: satellite/satellitedb/attribution_test.go [hunk 1]
```diff
@@ -39,7 +40,10 @@ func TestUsers(t *testing.T) {
 		require.NoError(t, err)
 
 		// create an user with no partnerID
+		userID, err := uuid.New()
```
Comment (mobyvb) [inline satellite/satellitedb/attribution_test.go]: no need for this if you are creating with `testrand.UUID` below

File: satellite/satellitedb/users.go [hunk 2]
```diff
@@ -48,9 +48,9 @@ func (users *users) GetByEmail(ctx context.Context, email string) (_ *console.Us
 // Insert is a method for inserting user into the database
 func (users *users) Insert(ctx context.Context, user *console.User) (_ *console.User, err error) {
 	defer mon.Task()(&ctx)(&err)
-	userID, err := uuid.New()
-	if err != nil {
-		return nil, err
+
+	if user.ID.IsZero() {
```
Comment (navillasa) [inline satellite/satellitedb/users.go]: do other fields need to be checked for nil values?

Comment (VinozzZ) [reply satellite/satellitedb/users.go]: I think the caller checks them atm

File: satellite/console/service.go [hunk 3]
```diff
@@ -335,7 +335,11 @@ func (s *Service) CreateUser(ctx context.Context, user CreateUser, tokenSecret R
 	}
 
 	err = withTx(tx, func(tx DBTx) error {
+		if err != nil {
```
Comment (mobyvb) [inline satellite/console/service.go]: why is this error check here?

Comment (VinozzZ) [reply satellite/console/service.go]: oops

**Comment by mobyvb** [comment]
Looks good but I'll hold off on approving until someone from the Kiev team has a chance to take a look.

**Comment by mobyvb** [comment]
@rikysya if we keep user ID creation in the DB layer, it will mean we cannot redeem a referral token in one step. We would have to verify and reserve the token, then create the user on the satellite, then finally register the new user's ID with the token on the referral manager.
If we create the user ID outside of the DB layer, we can validate and redeem the referral token in one step, then create the user on the satellite. Then we do not have to worry about the token getting stuck in the middle of validation and redemption if there is some error creating the user.

I am not sure what a better solution is, but it would be nice if we could avoid two separate requests to the referral manager.

File: satellite/satellitedb/users.go (hunk 2 above)
Comment (rikysya) [reply satellite/satellitedb/users.go]: that's is not enough to check if it's valid
`func New() (*UUID, error) {
	uuid := new(UUID)

	n, err := io.ReadFull(rand.Reader, uuid[:])
	if err != nil {
		return nil, err
	} else if n != len(uuid) {
		return nil, errors.New(fmt.Sprintf("insufficient random data (expected: %d, read: %d)", len(uuid), n))
	} else {
		// variant bits; for more info
		// see https://www.ietf.org/rfc/rfc4122.txt section 4.1.1
		uuid[8] = uuid[8]&0x3f | 0x80
		// version 4 (pseudo-random); for more info
		// see https://www.ietf.org/rfc/rfc4122.txt section 4.1.3
		uuid[6] = uuid[6]&0x0f | 0x40
	}

	return uuid, nil
}
`

File: satellite/console/consoleweb/consoleql/mutation.go [hunk 4]
```diff
@@ -86,6 +86,11 @@ func rootMutation(log *zap.Logger, service *console.Service, mailService *mailse
 					refUserID, _ := p.Args[ReferrerUserID].(string)
 
 					createUser := fromMapCreateUser(input)
+					userID, err := uuid.New()
```
Comment (crawter) [inline satellite/console/consoleweb/consoleql/mutation.go]: i would prefer to create userID in Service or in repository

Comment (crawter) [reply satellite/console/consoleweb/consoleql/mutation.go]: api should not change flow of business logic

File: satellite/satellitedb/users.go (hunk 2 above)
Comment (mobyvb) [reply satellite/satellitedb/users.go]: If we implement the random generation at a lower level (in `CreateUser` in `service.go`), I don't think randomness validation is necessary since we will be sure of how it is being generated.

File: satellite/console/projectmembers_test.go [hunk 5]
```diff
@@ -160,31 +160,38 @@ func TestProjectMembersRepository(t *testing.T) {
 
 func prepareUsersAndProjects(ctx context.Context, t *testing.T, users console.Users, projects console.Projects) ([]*console.User, []*console.Project) {
 	usersList := []*console.User{{
+		ID: testrand.UUID(),
+
```
Comment (rikysya) [inline satellite/console/projectmembers_test.go]: remove empty line

File: satellite/console/usercredits_test.go [hunk 6]
```diff
@@ -292,6 +292,8 @@ func setupData(ctx context.Context, t *testing.T, db satellite.DB) (user *consol
 
 	// create an user
 	user, err = consoleDB.Users().Insert(ctx, &console.User{
+		ID: testrand.UUID(),
+
```
Comment (rikysya) [inline satellite/console/usercredits_test.go]: remove empty line

File: satellite/console/service.go [hunk 7]
```diff
@@ -364,12 +370,12 @@ func (s *Service) CreateUser(ctx context.Context, user CreateUser, tokenSecret R
 		if currentReward != nil {
 			var refID *uuid.UUID
 			if refUserID != "" {
-				refID, err = uuid.Parse(refUserID)
+				refID, err = uuid.Parse(refID)
```
Comment (mobyvb) [inline satellite/console/service.go]: This change looks like a mistake to me - shouldn't it still be `refUserID`?

File: satellite/console/service.go [hunk 8]
```diff
@@ -364,12 +370,12 @@ func (s *Service) CreateUser(ctx context.Context, user CreateUser, tokenSecret R
...
 				if err != nil {
 					return Error.Wrap(err)
 				}
 			}
-			newCredit, err := NewCredit(currentReward, Invitee, u.ID, refID)
+			newCredit, err := NewCredit(currentReward, Invitee, u.ID, refToken)
```
Comment (mobyvb) [inline satellite/console/service.go]: where does `refToken` come from? Why this change?
